### PR TITLE
Modified task status flow 

### PR DESCRIPTION
**Pull request by skv93-coder**
Date: 14 Mar 24

Developer Name: @skv93-coder 

---

## Issue Ticket Number

https://github.com/Real-Dev-Squad/website-my/issues/554

## Description

This PR adds restrictions for task status change according to this issue ticket https://github.com/Real-Dev-Squad/website-my/issues/554
### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [x] Yes
- [ ] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
[Screencast from 14-03-24 06:55:19 AM IST.webm](https://github.com/Real-Dev-Squad/website-backend/assets/68867418/0bf60156-97bb-4564-a3f5-dc72db25a843)

<details>
<summary>Screenshot 1</summary>

</details>

## Test Coverage
Unit tests
Screenshot from 2024-03-14 07-41-58
Integration tests
Screenshot from 2024-03-14 07-40-00

<details>
<summary>Screenshot 1</summary>

</details>

### CODE REVIEW

File: test/integration/tasks.test.js
Comment (tejaskh3) [inline test/integration/tasks.test.js:1]: Can we combine the cases like describe("Should give 400") and then combine the cases or is it fine this way too?

**Review by iamitprakash — CHANGES_REQUESTED** [review]
can we have better PR title

**Comment by skv93-coder** [comment]
> can we have better PR title

Done

File: controllers/tasks.js [hunk 1]
```diff
@@ -347,6 +347,32 @@ const updateTaskStatus = async (req, res, next) => {
           return res.boom.badRequest("Status cannot be updated. Task is not done yet");
```
Comment (iamitprakash) [inline controllers/tasks.js:347]: can we have better message

File: controllers/tasks.js [hunk 2]
```diff
@@ -347,6 +347,32 @@ const updateTaskStatus = async (req, res, next) => {
...
+        const isCurrentTaskStatusBlock = task.taskData.status === TASK_STATUS.BLOCKED;
+        const isNewTaskStatusInProgress = status === TASK_STATUS.IN_PROGRESS;
+        const isNewTaskStatusBlock = status === TASK_STATUS.BLOCKED;
+        const isCurrProgress100 = parseInt(task.taskData.percentCompleted || 0) === 100;
+        const isCurrProgress0 = parseInt(task.taskData.percentCompleted || 0) === 0;
+        const isNewProgress100 = !!req.body.percentCompleted && parseInt(req.body.percentCompleted) === 100;
+        const isNewProgress0 = !!req.body.percentCompleted !== undefined && parseInt(req.body.percentCompleted) === 0;
+        if (isCurrentTaskStatusInProgress && !isNewTaskStatusBlock && !isCurrProgress100 && !isNewProgress100) {
+          return res.boom.badRequest(
+            "The status of task can not be changed from In progress until progress of task is not 100%."
```
Comment (iamitprakash) [inline controllers/tasks.js:361]: what about changing from Assigned to Backlog

File: controllers/tasks.js [hunk 3]
```diff
@@ -347,6 +347,32 @@ const updateTaskStatus = async (req, res, next) => {
...
+          );
+        }
+
+        if (isCurrentTaskStatusBlock && !isNewTaskStatusInProgress && !isCurrProgress100 && !isNewProgress100) {
+          return res.boom.badRequest(
+            "The status of task can not be changed from Blocked until progress of task is not 100%."
+          );
+        }
+        if (isNewTaskStatusInProgress && !isCurrentTaskStatusBlock && !isCurrProgress0 && !isNewProgress0) {
+          return res.boom.badRequest(
```
Comment (iamitprakash) [inline controllers/tasks.js:371]: same here

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: controllers/tasks.js (hunk 2 above)
Comment (skv93-coder) [reply controllers/tasks.js:361]: This route is not used to update task status to the backlog. The admin can anytime update a task to Backlog as before.

File: controllers/tasks.js (hunk 1 above)
Comment (skv93-coder) [reply controllers/tasks.js:347]: Updated the msg in PR https://github.com/Real-Dev-Squad/website-backend/pull/1937

File: controllers/tasks.js (hunk 3 above)
Comment (skv93-coder) [reply controllers/tasks.js:374]: Same as above this controller is not used for updating task to backlog status

**Review by iamitprakash — APPROVED** [review]

**Review by Achintya-Chatterjee — DISMISSED** [review]

**Review by iamitprakash — APPROVED** [review]

**Pull request merged**